src/network.py

```diff
@@ -0,0 +1,179 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Abstract the network configuration for the UPF service."""
+import ipaddress
+import logging
+import re
+from typing import List
+
+from machine import ExecError, Machine
+
+logger = logging.getLogger(__name__)
+
+
+class Network:
+    """Abstract the network configuration for the UPF service."""
+
```

Comment:
I am not sure about the way this class is implemented. I agree with having a class that handles the network configuration, but it seems to be acting at different abstraction layers, from the high level configuration, to the low-level interaction with command outputs.

I think splitting this a bit further would also make the tests a bit easier. I would for example highly consider using a library like `pyroute2` to handle the lower level access to interfaces, IP addresses and routes. Similarly, I would consider a library like `python-iptables` to handle the firewall rules.

If we do not want to bring 3rd party libraries as dependencies, I would suggest at least extracting those things to other separate classes. Maybe something like:

```python
nic = Interface('access')
if not nic.exists():
    logger.error("doh")
ip = nic.ip
```

Smaller objects like that would be easier to test, and the testing the higher level logic of testing the network configuration of the whole UPF would need less setup.

What do you think?

Reply:
I went ahead and did that, let me know what you think